source/Fleet.cpp

```diff
@@ -496,45 +531,33 @@ void Fleet::Place(const System &system, Ship &ship)
 int64_t Fleet::Strength() const
 {
 	int64_t sum = 0;
-	for(const Variant &variant : variants)
-	{
-		int64_t thisSum = 0;
-		for(const Ship *ship : variant.ships)
-			thisSum += ship->Cost();
-		sum += thisSum * variant.weight;
-	}
+	for(auto &variant : variants)
+		sum += variant.first.Strength() * variant.second;
+	for(auto &variant : stockVariants)
+		sum += variant.first->Strength() * variant.second;
 	return sum / total;
 }
 
 
 
-Fleet::Variant::Variant(const DataNode &node)
-{
-	weight = 1;
-	if(node.Token(0) == "variant" && node.Size() >= 2)
-		weight = node.Value(1);
-	else if(node.Token(0) == "add" && node.Size() >= 3)
-		weight = node.Value(2);
-	
-	for(const DataNode &child : node)
-	{
-		int n = 1;
-		if(child.Size() >= 2 && child.Value(1) >= 1.)
-			n = child.Value(1);
-		ships.insert(ships.end(), n, GameData::Ships().Get(child.Token(0)));
-	}
-}
-
-
-
-const Fleet::Variant &Fleet::ChooseVariant() const
+const Variant &Fleet::ChooseVariant() const
 {
 	// Pick a random variant based on the weights.
 	unsigned index = 0;
-	for(int choice = Random::Int(total); choice >= variants[index].weight; ++index)
-		choice -= variants[index].weight;
-	
-	return variants[index];
+	// Choose between stockVariants and variants.
+	int chosen = Random::Int(total);
+	if(chosen < stockTotal)
+	{
+		for(int choice = Random::Int(stockTotal); choice >= stockVariants[index].second; ++index)
+			choice -= stockVariants[index].second;
+		return *stockVariants[index].first;
+	}
+	else
+	{
+		for(int choice = Random::Int(total - stockTotal); choice >= variants[index].second; ++index)
```

Comment:
What happens with empty variants? Does this code still work if variants has zero elements (same for stockVariants above)?

Reply:
If `variants` is empty, then `total == stockTotal`, so `if(chosen < stockTotal)` will always be true. If `stockVariants` is empty, then the reverse is true where `total == total - stockTotal` and `if(chosen < stockTotal)` will always return false. If both `variants` and `stockVariants` are empty then we don't even have a fleet.

Reply:
I wasn't just talking about the normal cases. What if somebody makes a mistake in mission files and uses multiple remove of variants to get to the situation where we do have a `fleet` where both `variants` and `stockVariants` are empty?

Are we really sure that such situation cannot happen? And if we are not sure and it does happen, would the code then just result in placing an empty fleet or could we expect some sort of crash then? (The empty fleet placement in such case would be okay for me, the game crashing would not be okay for me.)

Reply:
Looks like Fleet::Enter and Fleet::Place already handle the case of having no variants or having variants with no ships by returning if there are no variants or if the variant that is chosen returns no ships.